lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -383,20 +491,7 @@ OpFoldResult AndPrimOp::fold(ArrayRef<Attribute> operands) {
 }
 
 OpFoldResult OrPrimOp::fold(ArrayRef<Attribute> operands) {
-  // or(x, invalid) -> x
-  // or(invalid, x) -> x
-  //
-  // This is legal because it aligns with the Scala FIRRTL Compiler
-  // interpretation of lowering invalid to constant zero before constant
-  // propagation.
-  if (operands[0].dyn_cast_or_null<InvalidValueAttr>() &&
-      rhs().getType() == getType())
-    return rhs();
-  if (operands[1].dyn_cast_or_null<InvalidValueAttr>() &&
-      lhs().getType() == getType())
-    return lhs();
-
-  if (auto rhsCst = operands[1].dyn_cast_or_null<IntegerAttr>()) {
+  if (auto rhsCst = getConstant(operands[1])) {
```

Comment:
This now only has to check the RHS because `or(invalid, x)` is properly canonicalized to `or(x, invalid)` because invalid is zero-like?

Reply:
Yeah exactly. The `OrPrimOp` is marked as commutative, and the `InvalidValueOp` folds to a `InvalidValueAttr` (which I guess MLIR interprets as "the constant"), so a builtin pattern will swap the operands on that `OrPrimOp`.